sdk/python/tests/test_generate_protos.py

```
@@ -0,0 +1,133 @@
+        with pytest.raises(generator.CodegenError, match="required codegen tool is missing"):
+            generator.verify_tool_versions()
+
+
+def test_tool_versions_fail_closed_on_pin_mismatch() -> None:
+    installed: Mapping[str, str] = {
+        "grpcio": "1.76.0",
+        "grpcio-tools": "1.77.0",
+        "protobuf": "6.31.1",
+    }
+    with patch.object(generator.metadata, "version", side_effect=installed.__getitem__):
+        with pytest.raises(generator.CodegenError, match="grpcio-tools.*1.77.0.*1.76.0"):
```

Ruff (0.15.21)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)